File: sdk/typescript/test/helpers/harness.ts
import { mkdtemp, rm } from "node:fs/promises";
import { existsSync } from "node:fs";
import os from "node:os";
import path from "node:path";
import readline from "node:readline";
import { spawn, type ChildProcessByStdio } from "node:child_process";
import type { Readable } from "node:stream";
import { fileURLToPath } from "node:url";

const repoRoot =
  process.env.OPENYAK_REPO_ROOT ??
  path.resolve(fileURLToPath(new URL("../../../../", import.meta.url)));
const rustRoot = path.join(repoRoot, "rust");

interface ProcessCommand {
  command: string;
  args: string[];
}

type ManagedChildProcess = ChildProcessByStdio<null, Readable, Readable>;

interface ManagedProcess {
  child: ManagedChildProcess;
  close(): Promise<void>;
}

export interface MockAnthropicHarness extends ManagedProcess {
  baseUrl: string;
}

export interface OpenyakServerHarness extends ManagedProcess {
  baseUrl: string;
  workspace: string;
}

export async function startMockAnthropicService(): Promise<MockAnthropicHarness> {
  const proc = await startProcess(
    resolveStandaloneRustBinary(
      "mock-anthropic-service",
      "MOCK_ANTHROPIC_SERVICE_BIN",
    ),
    {},
    /^MOCK_ANTHROPIC_BASE_URL=(.+)$/,
  );
  const baseUrl = proc.match[1];
  if (!baseUrl) {
    throw new Error("mock-anthropic-service did not report a base URL");
  }
  return {
    ...proc,
    baseUrl,
  };
}

export async function startOpenyakServer(
  env: NodeJS.ProcessEnv,
): Promise<OpenyakServerHarness> {
  const workspace = await mkdtemp(path.join(os.tmpdir(), "openyak-sdk-alpha-"));
  const proc = await startProcess(
    resolveOpenyakServerCommand(),
    {
      cwd: workspace,
      env,
    },
    /^Local thread server listening on (http:\/\/.+)$/,
  );
  const baseUrl = proc.match[1];
  if (!baseUrl) {
    throw new Error("openyak server did not report a base URL");
  }
  return {
    ...proc,
    baseUrl,
    workspace,
    async close() {
      await proc.close();
      await rm(workspace, { recursive: true, force: true });
    },
  };
}

export async function withServerHarness<T>(
  fn: (harness: { mock: MockAnthropicHarness; server: OpenyakServerHarness }) => Promise<T>,
): Promise<T> {
  const mock = await startMockAnthropicService();
  const server = await startOpenyakServer({
    ...process.env,
    ANTHROPIC_API_KEY: "test-sdk-key",
    ANTHROPIC_BASE_URL: mock.baseUrl,
  });
  try {
    return await fn({ mock, server });
  } finally {
    await server.close();
    await mock.close();
  }
}

function resolveStandaloneRustBinary(name: string, envVar: string): ProcessCommand {
  const override = process.env[envVar];
  if (override) {
    return { command: override, args: [] };
  }

  const binaryName = process.platform === "win32" ? `${name}.exe` : name;
  const builtBinary = path.join(rustRoot, "target", "debug", binaryName);
  if (existsSync(builtBinary)) {
    return { command: builtBinary, args: [] };
  }

  return {
    command: process.env.CARGO ?? "cargo",
    args: [
      "run",
      "--manifest-path",
      path.join(rustRoot, "Cargo.toml"),
      "--quiet",
      "--bin",
      name,
      "--",
    ],
  };
}

function resolveOpenyakServerCommand(): ProcessCommand {
  const override = process.env.OPENYAK_SERVER_BIN;
  if (override) {
    return { command: override, args: ["server"] };
  }

  const binaryName = process.platform === "win32" ? "openyak.exe" : "openyak";
  const builtBinary = path.join(rustRoot, "target", "debug", binaryName);
  if (existsSync(builtBinary)) {
    return { command: builtBinary, args: ["server"] };
  }

  return {
    command: process.env.CARGO ?? "cargo",
    args: [
      "run",
      "--manifest-path",
      path.join(rustRoot, "Cargo.toml"),
      "--quiet",
      "--bin",
      "openyak",
      "--",
      "server",
    ],
  };
}

async function startProcess(
  proc: ProcessCommand,
  options: {
    cwd?: string;
    env?: NodeJS.ProcessEnv;
  },
  matcher: RegExp,
): Promise<ManagedProcess & { match: RegExpMatchArray }> {
  const child = spawn(proc.command, proc.args.concat(["--bind", "127.0.0.1:0"]), {
    cwd: options.cwd ?? repoRoot,
    env: options.env ?? process.env,
    stdio: ["ignore", "pipe", "pipe"],
  });
  const match = await waitForMatch(child, matcher, 90_000);
  return {
    child,
    match,
    async close() {
      await terminate(child);
    },
  };
}

function waitForMatch(
  child: ManagedChildProcess,
  matcher: RegExp,
  timeoutMs: number,
): Promise<RegExpMatchArray> {
  return new Promise((resolve, reject) => {
    const stderrChunks: string[] = [];
    child.stderr.on("data", (chunk: Buffer | string) => {
      stderrChunks.push(String(chunk));
    });

    const timer = setTimeout(() => {
      cleanup();
      reject(
        new Error(
          `process did not emit startup line within ${timeoutMs}ms: ${stderrChunks.join("")}`,
        ),
      );
    }, timeoutMs);

    const rl = readline.createInterface({ input: child.stdout });

    const onExit = (code: number | null, signal: NodeJS.Signals | null) => {
      cleanup();
      reject(
        new Error(
          `process exited before startup line (code=${code}, signal=${signal}): ${stderrChunks.join("")}`,
        ),
      );
    };

    const onLine = (line: string) => {
      const match = line.match(matcher);
      if (!match) {
        return;
      }
      cleanup();
      resolve(match);
    };

    const cleanup = () => {
      clearTimeout(timer);
      rl.close();
      child.off("exit", onExit);
      rl.off("line", onLine);
    };

    child.on("exit", onExit);
    rl.on("line", onLine);
  });
}

async function terminate(child: ManagedChildProcess): Promise<void> {
  if (child.exitCode !== null) {
    return;
  }
  child.kill();
  await new Promise<void>((resolve) => {
    const timer = setTimeout(() => {
      if (child.exitCode === null) {
        child.kill("SIGKILL");
      }
      resolve();
    }, 5_000);
    child.once("exit", () => {
      clearTimeout(timer);
      resolve();
    });
  });
}

File: sdk/python/tests/helpers/harness.py
from __future__ import annotations

import os
import queue
import re
import shutil
import subprocess
import tempfile
import threading
import time
from collections.abc import Iterator
from contextlib import contextmanager
from dataclasses import dataclass
from pathlib import Path

REPO_ROOT = Path(os.environ.get("OPENYAK_REPO_ROOT", Path(__file__).resolve().parents[4]))
RUST_ROOT = REPO_ROOT / "rust"


@dataclass(slots=True)
class ManagedProcess:
    process: subprocess.Popen[str]
    _stderr_chunks: list[str]

    def close(self) -> None:
        if self.process.poll() is not None:
            return
        self.process.terminate()
        try:
            self.process.wait(timeout=5)
        except subprocess.TimeoutExpired:
            self.process.kill()
            self.process.wait(timeout=5)

    @property
    def stderr_text(self) -> str:
        return "".join(self._stderr_chunks)


@dataclass(slots=True)
class MockAnthropicHarness:
    base_url: str
    managed: ManagedProcess

    def close(self) -> None:
        self.managed.close()


@dataclass(slots=True)
class OpenyakServerHarness:
    base_url: str
    workspace: Path
    managed: ManagedProcess
    cleanup_workspace: bool = True

    def close(self) -> None:
        try:
            self.managed.close()
        finally:
            if self.cleanup_workspace:
                shutil.rmtree(self.workspace, ignore_errors=True)


@dataclass(slots=True)
class ServerHarness:
    mock: MockAnthropicHarness
    server: OpenyakServerHarness

    def close(self) -> None:
        try:
            self.server.close()
        finally:
            self.mock.close()


@contextmanager
def server_harness() -> Iterator[ServerHarness]:
    mock = start_mock_anthropic_service()
    server = start_openyak_server(
        {
            **os.environ,
            "ANTHROPIC_API_KEY": "test-sdk-key",
            "ANTHROPIC_BASE_URL": mock.base_url,
        }
    )
    harness = ServerHarness(mock=mock, server=server)
    try:
        yield harness
    finally:
        harness.close()


def start_mock_anthropic_service() -> MockAnthropicHarness:
    managed, match = _start_process(
        _resolve_standalone_rust_binary(
            "mock-anthropic-service",
            "MOCK_ANTHROPIC_SERVICE_BIN",
        ),
        {},
        re.compile(r"^MOCK_ANTHROPIC_BASE_URL=(.+)$"),
    )
    return MockAnthropicHarness(base_url=match.group(1), managed=managed)


def start_openyak_server(env: dict[str, str]) -> OpenyakServerHarness:
    workspace = Path(tempfile.mkdtemp(prefix="openyak-python-sdk-alpha-"))
    return start_openyak_server_in(workspace, env, cleanup_workspace=True)


def start_openyak_server_in(
    workspace: Path,
    env: dict[str, str],
    *,
    bind: str = "127.0.0.1:0",
    cleanup_workspace: bool = False,
) -> OpenyakServerHarness:
    managed, match = _start_process(
        _resolve_openyak_server_command(),
        {
            "cwd": workspace,
            "env": env,
            "bind": bind,
        },
        re.compile(r"^Local thread server listening on (http://.+)$"),
    )
    return OpenyakServerHarness(
        base_url=match.group(1),
        workspace=workspace,
        managed=managed,
        cleanup_workspace=cleanup_workspace,
    )


def _resolve_standalone_rust_binary(name: str, env_var: str) -> list[str]:
    override = os.environ.get(env_var)
    if override:
        return [override]

    binary_name = f"{name}.exe" if os.name == "nt" else name
    built_binary = RUST_ROOT / "target" / "debug" / binary_name
    if built_binary.exists():
        return [str(built_binary)]

    cargo = os.environ.get("CARGO", "cargo")
    return [
        cargo,
        "run",
        "--manifest-path",
        str(RUST_ROOT / "Cargo.toml"),
        "--quiet",
        "--bin",
        name,
        "--",
    ]


def _resolve_openyak_server_command() -> list[str]:
    override = os.environ.get("OPENYAK_SERVER_BIN")
    if override:
        return [override, "server"]

    binary_name = "openyak.exe" if os.name == "nt" else "openyak"
    built_binary = RUST_ROOT / "target" / "debug" / binary_name
    if built_binary.exists():
        return [str(built_binary), "server"]

    cargo = os.environ.get("CARGO", "cargo")
    return [
        cargo,
        "run",
        "--manifest-path",
        str(RUST_ROOT / "Cargo.toml"),
        "--quiet",
        "--bin",
        "openyak",
        "--",
        "server",
    ]


def _start_process(
    command: list[str],
    options: dict[str, object],
    matcher: re.Pattern[str],
) -> tuple[ManagedProcess, re.Match[str]]:
    process = subprocess.Popen(
        command + ["--bind", str(options.get("bind", "127.0.0.1:0"))],
        cwd=str(options.get("cwd", REPO_ROOT)),
        env=options.get("env", os.environ),
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        text=True,
        encoding="utf-8",
    )
    stdout_queue: queue.Queue[str] = queue.Queue()
    stderr_chunks: list[str] = []
    threading.Thread(
        target=_pump_stdout,
        args=(process, stdout_queue),
        daemon=True,
    ).start()
    threading.Thread(
        target=_pump_stderr,
        args=(process, stderr_chunks),
        daemon=True,
    ).start()

    deadline = time.monotonic() + 90
    while time.monotonic() < deadline:
        if process.poll() is not None:
            raise RuntimeError(
                "process exited before startup line "
                f"(code={process.returncode}): {''.join(stderr_chunks)}"
            )
        try:
            line = stdout_queue.get(timeout=0.1)
        except queue.Empty:
            continue
        match = matcher.match(line)
        if match:
            return ManagedProcess(process=process, _stderr_chunks=stderr_chunks), match

    process.terminate()
    raise RuntimeError(
        f"process did not emit startup line within 90s: {''.join(stderr_chunks)}"
    )


def _pump_stdout(process: subprocess.Popen[str], sink: queue.Queue[str]) -> None:
    assert process.stdout is not None
    for line in process.stdout:
        sink.put(line.rstrip("\r\n"))


def _pump_stderr(process: subprocess.Popen[str], sink: list[str]) -> None:
    assert process.stderr is not None
    for line in process.stderr:
        sink.append(line)
